### PR TITLE
[Windows] Improve win32 handling of output pipes

### LINKED ISSUE #13964: (Windows) Cannot Redirect the access_log output to (stderr/stdout/null)
### description 

On Windows there is no easy way to redirect access log output to standard output/error. 

original description:

> while documenting basic logging (#13927) it became apparent that some of the ~default (container) configs might not work on > windows, and this pointed to an underlying issue in terms of handling those paths and/or documenting something similar
>
> one suggestion that came up  was to have some kind of config identifier "$dev_stdout" eg, that would #justwork on both

### PR DESCRIPTION
**Pull request by davinci26**
Commit Message:
Signed-off-by: Sotiris Nanopoulos <sonanopo@microsoft.com>

Fixes #13964

Additional Description:
Risk Level: Low
Testing: Unit testing
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by davinci26** [comment]
@envoyproxy/windows-dev please review. Note that this item is marked as a beta item

File: source/common/filesystem/win32/filesystem_impl.cc [hunk 1]
```diff
@@ -31,8 +60,37 @@ Api::IoCallBoolResult FileImplWin32::open(FlagSet in) {
     return resultSuccess(true);
   }
 
-  auto flags = translateFlag(in);
-  fd_ = CreateFileA(path_.c_str(), flags.access_, FILE_SHARE_READ | FILE_SHARE_WRITE, 0,
+  auto std_handle = toStandardHandle(path_);
```
Comment (sunjayBhatia) [inline source/common/filesystem/win32/filesystem_impl.cc:63]: nit: maybe just add some comments in this section to ensure its easy to understand with a quick glance, should already be pretty self explanatory but for those that are not familiar

**Review by sunjayBhatia — CHANGES_REQUESTED** [review]
Just a couple nits in comments

You could also maybe add a file access log integration test to ensure stdout/err does work since that is what this change is attempting to support, but I'm not sure if we have good infrastructure already for testing that?

Another place that could be an alternative to where this is placed would be to be in the file access log implementation itself, but given that is the only place that uses `open` this seems like a better platform specific place to put it. If we somehow have other code that uses `open` in the future, might have to rethink it maybe

**Comment by davinci26** [comment]
I will look into if I could add integration tests in access_logger although afaik it only had unit tests

> Another place that could be an alternative to where this is placed would be to be in the file access log implementation itself, but given that is the only place that uses open this seems like a better platform specific place to put it.

Right now the FileImpl API does not offer a `read` api only `write`. I feel that in any case and even for further usecases the `fileImpl` implementation should do the translation and support the output streams.

File: source/common/filesystem/win32/filesystem_impl.cc [hunk 2]
```diff
@@ -19,6 +19,35 @@
 namespace Envoy {
 namespace Filesystem {
 
+static const absl::flat_hash_map<std::string, DWORD> unix_output_pipes_to_windows_standard_handle{
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/common/filesystem/win32/filesystem_impl.cc:23]: No non-POD statics please. See https://github.com/envoyproxy/envoy/blob/master/STYLE.md :
> The Google C++ style guide points out that non-PoD static and global variables are forbidden. This includes types such as std::string. We encourage the use of the advice in the C++ FAQ on the static initialization fiasco for how to best handle this.

File: source/common/filesystem/win32/filesystem_impl.cc (hunk 1 above)
Comment (jmarantz) [reply source/common/filesystem/win32/filesystem_impl.cc:70]: please use an explicit type here, also per style guide.

File: source/common/filesystem/win32/filesystem_impl.cc [hunk 3]
```diff
@@ -19,6 +19,36 @@
...
+namespace {
+static const absl::flat_hash_map<std::string, DWORD> unix_output_pipes_to_windows_standard_handle{
+    {"/dev/stdin", STD_INPUT_HANDLE},       {"/dev/fd/0", STD_INPUT_HANDLE},
+    {"/proc/self/fd/0", STD_INPUT_HANDLE},
+
+    {"/dev/stdout", STD_OUTPUT_HANDLE},     {"/dev/fd/1", STD_OUTPUT_HANDLE},
+    {"/proc/self/fd/1", STD_OUTPUT_HANDLE},
+
+    {"/dev/stderr", STD_ERROR_HANDLE},      {"/dev/fd/2", STD_ERROR_HANDLE},
+    {"/proc/self/fd/2", STD_ERROR_HANDLE}};
```
Comment (jmarantz) [inline source/common/filesystem/win32/filesystem_impl.cc:31]: rather than this map, why don't we just add new APIs to the FileSystem class that return already-opened stdio handles. No need to make extra OS calls for this.

Also, why do we need to open new handles to stdio at all in the proxy?

Comment (davinci26) [reply source/common/filesystem/win32/filesystem_impl.cc:31]: >  why don't we just add new APIs to the FileSystem class that return already-opened stdio handles. No need to make extra OS calls for this.

Can you explain a bit more what you mean here?

Do you mean a function added in the `FileSystem` API  like  `handle_t getStandardOutputHandle()`?

> Also, why do we need to open new handles to stdio at all in the proxy?

We dont open new handles, but for the standard handles the OS is the owner of those handle. So we are asking the OS for the latest handle and the OS increments the ref count.

Comment (jmarantz) [reply source/common/filesystem/win32/filesystem_impl.cc:31]: > Do you mean a function added in the FileSystem API like handle_t getStandardOutputHandle()?

Yeah, to be more specific, see `include/envoy/filesystem/filesystem.h`. My suggestion would be to add:
```
FIle& standardInput() PURE;
FIle& standardOutput() PURE;
```
to class `Instance`. It would return those descriptors in an open state, and the reference would be owned by `Instance`. That's my proposal, anyway. I think the implementation of each of these would be trivial in both windows and posix.

I'm not sure if this captures what usage model you have in mind. What usage model do you have in mind for these, anyway?

Comment (sunjayBhatia) [reply source/common/filesystem/win32/filesystem_impl.cc:31]: cross linking in case anyone else needs it: https://github.com/envoyproxy/envoy/issues/13964

tldr: support file access logging to stdout/err to emulate Linux/macos etc.

**Comment by davinci26** [comment]
We are considering a new approach. See discussion in #13964
